File: client/src/types/ae.ts
export type ApiNumericBinning = 'uniform' | 'quintile' | 'custom';

export interface ApiAeXVariableNumeric {
    kind: 'numeric';
    name: string;
    binning: ApiNumericBinning;
    bin_count?: number | null;
    custom_edges?: number[] | null;
}

export interface ApiAeXVariableDate {
    kind: 'date';
    name: string;
    binning: ApiNumericBinning;
    bin_count?: number | null;
    custom_edges?: string[] | null;
}

export interface ApiCategoricalGroupDefinition {
    name: string;
    values: string[];
    x_position?: number | null;
}

export interface ApiAeXVariableCategorical {
    kind: 'categorical';
    name: string;
    grouping: 'all_unique' | 'custom';
    groups?: ApiCategoricalGroupDefinition[] | null;
    remaining_name?: string;
    remaining_position?: number | null;
}

export type ApiAeAtomicVariable =
    | ApiAeXVariableNumeric
    | ApiAeXVariableDate
    | ApiAeXVariableCategorical;

export interface ApiAeCrossGroupDefinition {
    name: string;
    a_any: boolean;
    a_values: string[];
    b_any: boolean;
    b_values: string[];
    x_position?: number | null;
}

export interface ApiAeXVariableCross {
    kind: 'cross';
    a_variable: ApiAeAtomicVariable;
    b_variable: ApiAeAtomicVariable;
    groups?: ApiAeCrossGroupDefinition[] | null;
    remaining_name?: string;
    remaining_position?: number | null;
}

export type ApiAeXVariable = ApiAeAtomicVariable | ApiAeXVariableCross;

export interface ApiColumnMapping {
    policy_number_column?: string | null;
    face_amount_column?: string | null;
    mac_column?: string | null;
    mec_column?: string | null;
    man_column?: string | null;
    men_column?: string | null;
    moc_column?: string | null;
    cola_m1_column?: string | null;
}

export interface ApiAeUnivariateParameters {
    dataset_name: string;
    x_variable: ApiAeXVariable;
    split_variable?: ApiAeXVariable | null;
    application_id_column?: string | null;
    column_mapping?: ApiColumnMapping | null;
    exclusions?: ApiAeExclusions | null;
    poly_fit?: ApiAePolynomialFitParameters | null;
}

export interface ApiAeUnivariateFromConfigParameters {
    config_id: string;
    x_variable: ApiAeXVariable;
    split_variable?: ApiAeXVariable | null;
    application_id_column?: string | null;
    exclusions?: ApiAeExclusions | null;
    poly_fit?: ApiAePolynomialFitParameters | null;
}

export interface ApiAeUnivariateRow {
    variable_group: string;
    avg_x?: number | null;
    x_coord?: number | null;
    sample_size: number;
    deaths: number;
    expected_count: number;
    actual_amount: number;
    expected_amount: number;
    exposure_count: number;
    total_face_amount: number;
    ae: number | null;
    ae_amount: number | null;
    ae_ci_lower?: number | null;
    ae_ci_upper?: number | null;
    ae_amount_ci_lower?: number | null;
    ae_amount_ci_upper?: number | null;
}

export interface ApiAeColaM1StackedRow {
    x_group: string;
    total_deaths: number;
    deaths_by_m1: Record<string, number>;
    total_amount: number;
    amounts_by_m1: Record<string, number>;
}

export interface ApiAeColaM1StackedResults {
    causes: string[];
    rows: ApiAeColaM1StackedRow[];
    total_deaths: number;
    total_amount: number;
}

export interface ApiAeUnivariateResults {
    rows: ApiAeUnivariateRow[];
    split_results?: ApiAeUnivariateSplitResults[] | null;
    poly_fit?: ApiAePolynomialFitResults | null;
    cola_m1_stacked?: ApiAeColaM1StackedResults | null;
}

export interface ApiAeUnivariateSplitResults {
    split_group: string;
    rows: ApiAeUnivariateRow[];
    poly_fit?: ApiAePolynomialFitResults | null;
    cola_m1_stacked?: ApiAeColaM1StackedResults | null;
}

export interface ApiAeExclusions {
    exclude_cola_m1: string[];
    exclude_cola_m2_by_m1: Record<string, string[]>;
}

export interface ApiAePolynomialFitParameters {
    degree: 1 | 2 | 3;
    weighted: boolean;
}

export interface ApiAePolynomialFitResults {
    degree: 1 | 2 | 3;
    weighted: boolean;
    coefficients: number[];
    r2?: number | null;
    fit_table: {
        columns: Array<{ name: string; format: 'number' | 'percent' }>;
        rows: Array<Array<number | null>>;
    };
}

export interface ApiAeVariableLabelsParameters {
    dataset_name: string;
    variable: ApiAeAtomicVariable;
}

export interface ApiAeVariableLabelsResults {
    labels: string[];
}

File: app/modules/mortality_ae/models/ae.py
from __future__ import annotations

from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.modules.mortality_ae.models.chart import ApiChartTable


class ApiNumericBinning(StrEnum):
    UNIFORM = "uniform"
    QUINTILE = "quintile"
    CUSTOM = "custom"


class ApiAeXVariableNumeric(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kind: Literal["numeric"] = "numeric"
    name: str
    binning: ApiNumericBinning = ApiNumericBinning.QUINTILE
    bin_count: int | None = Field(default=5, ge=2, le=20)
    custom_edges: list[float] | None = None


class ApiAeXVariableDate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kind: Literal["date"] = "date"
    name: str
    binning: ApiNumericBinning = ApiNumericBinning.QUINTILE
    bin_count: int | None = Field(default=5, ge=2, le=20)
    custom_edges: list[str] | None = None


class ApiCategoricalGroupDefinition(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    values: list[str]
    x_position: float | None = None


class ApiAeXVariableCategorical(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kind: Literal["categorical"] = "categorical"
    name: str
    grouping: Literal["all_unique", "custom"] = "all_unique"
    groups: list[ApiCategoricalGroupDefinition] | None = None
    remaining_name: str = "Remaining"
    remaining_position: float | None = None


ApiAeAtomicVariable = ApiAeXVariableNumeric | ApiAeXVariableDate | ApiAeXVariableCategorical


class ApiAeCrossGroupDefinition(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    a_any: bool = False
    a_values: list[str] = Field(default_factory=list)
    b_any: bool = False
    b_values: list[str] = Field(default_factory=list)
    x_position: float | None = None


class ApiAeXVariableCross(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kind: Literal["cross"] = "cross"
    a_variable: ApiAeAtomicVariable
    b_variable: ApiAeAtomicVariable
    groups: list[ApiAeCrossGroupDefinition] | None = None
    remaining_name: str = "Remaining"
    remaining_position: float | None = None


ApiAeXVariable = ApiAeAtomicVariable | ApiAeXVariableCross


class ApiAePolynomialFitParameters(BaseModel):
    model_config = ConfigDict(extra="forbid")

    degree: int = Field(default=1, ge=1, le=3)
    weighted: bool = False


class ApiAePolynomialFitResults(BaseModel):
    model_config = ConfigDict(extra="forbid")

    degree: int
    weighted: bool
    coefficients: list[float]
    r2: float | None = None
    fit_table: ApiChartTable


class ApiColumnMapping(BaseModel):
    model_config = ConfigDict(extra="forbid")

    policy_number_column: str | None = None
    face_amount_column: str | None = None
    mac_column: str | None = None
    mec_column: str | None = None
    man_column: str | None = None
    men_column: str | None = None
    moc_column: str | None = None
    cola_m1_column: str | None = None


class ApiAeExclusions(BaseModel):
    model_config = ConfigDict(extra="forbid")

    exclude_cola_m1: list[str] = Field(default_factory=list)
    exclude_cola_m2_by_m1: dict[str, list[str]] = Field(default_factory=dict)


class ApiAeUnivariateParameters(BaseModel):
    model_config = ConfigDict(extra="forbid")

    dataset_name: str = ""
    x_variable: ApiAeXVariable
    split_variable: ApiAeXVariable | None = None
    application_id_column: str | None = None
    column_mapping: ApiColumnMapping | None = None
    exclusions: ApiAeExclusions | None = None
    poly_fit: ApiAePolynomialFitParameters | None = None


class ApiAeUnivariateFromConfigParameters(BaseModel):
    model_config = ConfigDict(extra="forbid")

    config_id: str
    x_variable: ApiAeXVariable
    split_variable: ApiAeXVariable | None = None
    application_id_column: str | None = None
    exclusions: ApiAeExclusions | None = None
    poly_fit: ApiAePolynomialFitParameters | None = None


class ApiAeUnivariateRow(BaseModel):
    model_config = ConfigDict(extra="forbid")

    variable_group: str
    avg_x: float | None = None
    x_coord: float | None = None
    sample_size: int
    deaths: float
    expected_count: float = 0.0
    actual_amount: float = 0.0
    expected_amount: float = 0.0
    exposure_count: float = 0.0
    total_face_amount: float = 0.0
    ae: float | None
    ae_amount: float | None = None
    ae_ci_lower: float | None = None
    ae_ci_upper: float | None = None
    ae_amount_ci_lower: float | None = None
    ae_amount_ci_upper: float | None = None


class ApiAeColaM1StackedRow(BaseModel):
    model_config = ConfigDict(extra="forbid")

    x_group: str
    total_deaths: float
    deaths_by_m1: dict[str, float]
    total_amount: float
    amounts_by_m1: dict[str, float]


class ApiAeColaM1StackedResults(BaseModel):
    model_config = ConfigDict(extra="forbid")

    causes: list[str]
    rows: list[ApiAeColaM1StackedRow]
    total_deaths: float
    total_amount: float


class ApiAeUnivariateSplitResults(BaseModel):
    model_config = ConfigDict(extra="forbid")

    split_group: str
    rows: list[ApiAeUnivariateRow]
    poly_fit: ApiAePolynomialFitResults | None = None
    cola_m1_stacked: ApiAeColaM1StackedResults | None = None


class ApiAeUnivariateResults(BaseModel):
    model_config = ConfigDict(extra="forbid")

    rows: list[ApiAeUnivariateRow]
    split_results: list[ApiAeUnivariateSplitResults] | None = None
    poly_fit: ApiAePolynomialFitResults | None = None
    cola_m1_stacked: ApiAeColaM1StackedResults | None = None
